### PR TITLE
Doc ieee and examples

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/boost/simd/function/bitfloating.hpp [hunk 1]
```diff
@@ -16,32 +16,32 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing bitfloating capabilities
+    This function object  transforms of an integer value to a floating point one
```
Comment (jfalcou) [inline include/boost/simd/function/bitfloating.hpp]: "This function object  transforms an integer value to a floating point one"

File: include/boost/simd/function/bitfloating.hpp [hunk 2]
```diff
@@ -16,32 +16,32 @@ namespace boost { namespace simd
...
+    with a preserving order formula. This function is the converse of @ref bitinteger.
 
-    Transform a pattern of bits stored in an integer value
-    in a floating point with different formulas according to
-    the integer @ref bitofsign value.
 
-    This function is the converse of @ref bitinteger
+    @par Note:
```
Comment (jfalcou) [inline include/boost/simd/function/bitfloating.hpp:24]: This is irrelevant : either its internal and shoudlnt be in the public API, or it's public and is documented normally.

File: include/boost/simd/function/bitfloating.hpp [hunk 3]
```diff
@@ -16,32 +16,32 @@ namespace boost { namespace simd
...
-    @see next, prev, successor,  predecessor,  nextafter, bitinteger
+       @snippet bitfloating.cpp bitfloating
+
+    @par Possible output:
+
+       @snippet bitfloating.txt bitfloating
 
   **/
-  as_floating_t<T> bitfloating(Value const & v0);
+  as_floating_t<Value> bitfloating(Value const & x);
```
Comment (jfalcou) [inline include/boost/simd/function/bitfloating.hpp]: Value shoud be IntegralValue

File: include/boost/simd/function/bitinteger.hpp [hunk 4]
```diff
@@ -16,32 +16,34 @@ namespace boost { namespace simd
...
     @ingroup group-ieee
-    Function object implementing bitinteger capabilities
-
-    Transforms a floating point value in a pattern of bits
-    stored in an integer with different formulas according to
-    the floating point @ref bitofsign.
-
+    This function object transforms a floating point value
+    in an integer with a preserving order formula.
     This function is the converse of @ref bitfloating.
```
Comment (jfalcou) [inline include/boost/simd/function/bitinteger.hpp:22]: "is the inverse"

File: include/boost/simd/function/bitinteger.hpp [hunk 5]
```diff
@@ -16,32 +16,34 @@ namespace boost { namespace simd
...
 
-    @par Semantic:
 
-    for any parameter @c x of floating type @c T
+    @par Note:
```
Comment (jfalcou) [inline include/boost/simd/function/bitinteger.hpp:25]: Again, irrelevant

File: include/boost/simd/function/bitinteger.hpp [hunk 6]
```diff
@@ -16,32 +16,34 @@ namespace boost { namespace simd
...
+    This is an internaly used utility function related to the computation of
+    floating successors or predecessors as the ordering of the integers obtained
+    by @c bifloating is the same as the original order on the floating elements,
+    an increment can be used to find the successor.
 
-    @code
-     as_integer_t<T> r = bitinteger(x);
-    @endcode
+    `next(x)` is simply `bitfloating(inc(bitinteger(x)))`
```
Comment (jfalcou) [inline include/boost/simd/function/bitinteger.hpp]: Same, no need for this here

File: include/boost/simd/function/bitofsign.hpp [hunk 7]
```diff
@@ -16,31 +16,22 @@ namespace boost { namespace simd
...
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing bitofsign capabilities
-
-    Returns a value of the same type as the input with all bits set to zero
+    This function object returns a value of the same type as
+    the input with all bits set to zero
     except the bit of sign which is preserved.
```
Comment (jfalcou) [inline include/boost/simd/function/bitofsign.hpp]: "except for the sign bit"

File: include/boost/simd/function/copysign.hpp [hunk 8]
```diff
@@ -16,25 +16,22 @@ namespace boost { namespace simd
...
  /*!
-
     @ingroup group-ieee
-    Function object implementing copysign capabilities
+    This function object returns @c x with the sign of @c y.
+
+   @par Note
 
-    Returns x with the sign of y
+     `copysign(x,y)` is similar to: `is_positive(y) ? abs(x) : -(abs(x))`
```
Comment (jfalcou) [inline include/boost/simd/function/copysign.hpp]: We have the example which is more useful than this note.

File: include/boost/simd/function/exponent.hpp [hunk 9]
```diff
@@ -16,37 +16,31 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing exponent capabilities
+    This function object returns the base (2) exponent of the floating input.
```
Comment (jfalcou) [inline include/boost/simd/function/exponent.hpp]: "the base 2" no need for ()

File: include/boost/simd/function/exponent.hpp [hunk 10]
```diff
@@ -16,37 +16,31 @@ namespace boost { namespace simd
...
 
-    For integral type inputs exponent is always 0 and mantissa reduces to identity.
+   @par Possible output:
 
-    @see mantissa,  frexp, ldexp
+     @snippet exponent.txt exponent
 
   **/
-  as_integer_t<Value> exponent(Value const & v0);
+  as_integer_t<Value> exponent(Value const & x);
```
Comment (jfalcou) [inline include/boost/simd/function/exponent.hpp]: Value should be RealValue

File: include/boost/simd/function/exponentbits.hpp [hunk 11]
```diff
@@ -16,23 +16,22 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing exponentbits capabilities
-
-    Returns the exponent bits of the floating input as an integer value,
+    This function object returns the exponent bits of the floating input as an integer value,
```
Comment (jfalcou) [inline include/boost/simd/function/exponentbits.hpp]: "of its arguments as an integer value"

File: include/boost/simd/function/exponentbits.hpp [hunk 12]
```diff
@@ -16,23 +16,22 @@ namespace boost { namespace simd
...
     the other bits (sign and mantissa) being just masked.
+    This function is not defined for integral types.
```
Comment (jfalcou) [inline include/boost/simd/function/exponentbits.hpp]: Remove

File: include/boost/simd/function/exponentbits.hpp [hunk 13]
```diff
@@ -16,23 +16,22 @@ namespace boost { namespace simd
...
+   @par Possible output:
 
-    @code
-    as_integer_t<T> r = exponentbits(x);
-    @endcode
+     @snippet exponentbits.txt exponentbits
 
   **/
-  as_integer_t<Value> exponentbits(Value const & v0);
+  as_integer_t<Value> exponentbits(Value const & x);
```
Comment (jfalcou) [inline include/boost/simd/function/exponentbits.hpp]: Value should be RealValue

File: include/boost/simd/function/fpclassify.hpp [hunk 14]
```diff
@@ -17,45 +17,45 @@ namespace boost { namespace simd
...
 
-    Categorizes floating point value arg into the following categories:
-    zero, subnormal, normal, infinite, NAN, or implementation-defined category.
-
-    @code
-    auto r = fpclassify(x);
-    @endcode
+    @par Notes
 
     This function is similar to std::fpclassify,  but the return type
```
Comment (jfalcou) [inline include/boost/simd/function/fpclassify.hpp]: I don't understand this sentence here

File: include/boost/simd/function/fpclassify.hpp [hunk 15]
```diff
@@ -17,45 +17,45 @@ namespace boost { namespace simd
...
+
+     @snippet fpclassify.cpp fpclassify
+
+   @par Possible output:
+
+     @snippet fpclassify.txt fpclassify
+
   **/
-  as_integer_t<Value> fpclassify(Value const & v0);
+  as_integer_t<Value> fpclassify(Value const & x);
```
Comment (jfalcou) [inline include/boost/simd/function/fpclassify.hpp]: Value -> RealValue

File: include/boost/simd/function/frexp.hpp [hunk 16]
```diff
@@ -16,15 +16,12 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing frexp capabilities
-
-    Computes a mantissa and an exponent pair for the input
+    This function object returns a mantissa and an exponent pair for the input
```
Comment (jfalcou) [inline include/boost/simd/function/frexp.hpp:20]: a mantissa or *the* mantissa ?

File: include/boost/simd/function/frexp.hpp [hunk 17]
```diff
@@ -16,15 +16,12 @@ namespace boost { namespace simd
...
 
     @par Semantic:
 
-    For every parameter of floating type @c T
+    For every parameter of floating type
```
Comment (jfalcou) [inline include/boost/simd/function/frexp.hpp:24]: the @c T must stay or the semantic definition is wrong

File: include/boost/simd/function/frexp.hpp [hunk 18]
```diff
@@ -33,38 +30,42 @@ namespace boost { namespace simd
...
-    @code
-    std:pair<T,T> p = frexp(x);
-    @endcode
-
-    can also be used.
-
     @par Note:
 
-    if you need integral type exponent (as in the standard library)  use @c ifrexp
+    Without the pedantic_ decorator  @ref Nan or @ref Inf are not handled properly.
```
Comment (jfalcou) [inline include/boost/simd/function/frexp.hpp]: Calling frexp on Nan or Inf is undefined behavior.

File: include/boost/simd/function/frexp.hpp [hunk 19]
```diff
@@ -33,38 +30,42 @@ namespace boost { namespace simd
...
 
-    This function splits a floating point value @c v f in a signed mantissa @c m and
-    an exponent @c e so that:  @f$v = m\times 2^e@f$,
-    with absolute value of @c m \f$\in [1/2, 1[\f$
+    if you need integral type exponent (as in the standard library)  use @ref ifrexp
```
Comment (jfalcou) [inline include/boost/simd/function/frexp.hpp:41]: Put that in a note at the end and add ifrexp to @seealso

File: include/boost/simd/function/frexp.hpp [hunk 20]
```diff
@@ -33,38 +30,42 @@ namespace boost { namespace simd
...
-    without pedantic_ the call is speedier, but special values as Nan or Inf are not handled properly.
-    std_ transmit the call to std::ifrexp. That implies that simd is ever emulated.
+   @par Possible output:
 
-    @see exponent, mantissa, copysign
+     @snippet frexp.txt frexp
 
   **/
-  std::pair<T, as_integer_t<Value>> frexp(Value const & v0);
+  std::pair<Value, Value> frexp(Value const & x);
```
Comment (jfalcou) [inline include/boost/simd/function/frexp.hpp]: Value -> RealValue

File: include/boost/simd/function/ifrexp.hpp [hunk 21]
```diff
@@ -16,11 +16,8 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing ifrexp capabilities
-
-    Computes a mantissa and an exponent pair for the input
+    This function object returns the Computes a mantissa and an exponent pair for the input
```
Comment (jfalcou) [inline include/boost/simd/function/ifrexp.hpp]: Sentence is all over the place and unreadable

File: include/boost/simd/function/ilogb.hpp [hunk 22]
```diff
@@ -16,29 +16,24 @@ namespace boost { namespace simd
...
+    This function object returns the integer truncation
     of the base 2 logarithm of x.
-
     It coincides with the @ref exponent function
     on all platforms supported.
 
-    @par Semantic:
+   @par Note
+
+     `ilogb(x)` is similar to: `saturated_(toint)(log2(x))`
```
Comment (jfalcou) [inline include/boost/simd/function/ilogb.hpp]: This is implementation detail, and is unneeded

File: include/boost/simd/function/ldexp.hpp [hunk 23]
```diff
@@ -16,35 +16,37 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing ldexp capabilities
-
-    The function multiply a floating entry \f$x\f$
+    This function object returns the The function multiply a floating entry \f$x\f$
```
Comment (jfalcou) [inline include/boost/simd/function/ldexp.hpp]: Unparsable sentence

File: include/boost/simd/function/mantissa.hpp [hunk 24]
```diff
@@ -16,32 +16,28 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing mantissa capabilities
+    This function object returns the signed mantissa of the floating input.
```
Comment (jfalcou) [inline include/boost/simd/function/mantissa.hpp]: of its flaoting point argument. 
Stay consistent and use argument everywhere

File: include/boost/simd/function/maxmag.hpp [hunk 25]
```diff
@@ -16,25 +16,16 @@ namespace boost { namespace simd
...
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing maxmag capabilities
-
-    Returns the input value which have the greatest absolute value.
-
-    @par Semantic:
+    This function object returns the input value which have the greatest absolute value.
```
Comment (jfalcou) [inline include/boost/simd/function/maxmag.hpp]: returns the value with the greatest absolute value ammong its arguments.

File: include/boost/simd/function/maxnum.hpp [hunk 26]
```diff
@@ -16,34 +16,22 @@ namespace boost { namespace simd
...
-    Function object implementing maxnum capabilities
-
-    Returns the maximum value, ignoring nans.
-
-    @par Semantic:
-
-    @code
-    auto r = maxnum(x,y);
-    @endcode
+    This function object returns the maximum value, ignoring nans.
```
Comment (jfalcou) [inline include/boost/simd/function/maxnum.hpp]: the greater value of its arguments, ignorign NaN.

look at std max wording.

File: include/boost/simd/function/maxnummag.hpp [hunk 27]
```diff
@@ -16,25 +16,17 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing maxnummag capabilities
-
-    Returns the input value which have the greatest absolute value, ignoring nan.
+    This function object returns the input value which have the greatest absolute value, ignoring nan.
```
Comment (jfalcou) [inline include/boost/simd/function/maxnummag.hpp]: see above for rephrasing

File: include/boost/simd/function/modf.hpp [hunk 28]
```diff
@@ -16,26 +16,24 @@ namespace boost { namespace simd
...
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing modf capabilities
+    This function object returns the Computes the integer and  fractional parts of the input
+
+   @par Note
+
+     `std::tie(t, f)= modf(x)` is similar to: ` t = trunc(x)` and `f = frac(x)`
```
Comment (jfalcou) [inline include/boost/simd/function/modf.hpp]: This is exactly waht a semantic block shoudl sya, not a mere Note. So removign it OR make it a semantic block

File: include/boost/simd/function/negate.hpp [hunk 29]
```diff
@@ -16,29 +16,23 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing negate capabilities
+    This function object returns  the first element multiplied by the @ref sign of the second.
```
Comment (jfalcou) [inline include/boost/simd/function/negate.hpp]: multiply by the sign is rather colloquial. 
I would say : "multiply by 1 or -1 depending on the fact its second argument is greater or lesser than zero."

File: include/boost/simd/function/next.hpp [hunk 30]
```diff
@@ -16,19 +16,9 @@ namespace boost { namespace simd
...
-
-    @par Semantic:
-
-    @code
-    auto r = next(x);
-    @endcode
-
-    @par Note
+    This function object returns the smallest representable value
+    strictly greater than the parameter with a few exceptions
```
Comment (jfalcou) [inline include/boost/simd/function/next.hpp]: which exception ...

File: include/boost/simd/function/safe_max.hpp [hunk 31]
```diff
@@ -16,31 +16,25 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing safe_max capabilities
-
-    Returns a safe_max relative to the input,  i.e. a
+    This function object returns a safe maximal value relative to the input,  i.e. a
```
Comment (jfalcou) [inline include/boost/simd/function/safe_max.hpp:20]: "to its argument"
"when multilied by it argument"

File: include/boost/simd/function/safe_max.hpp [hunk 32]
```diff
@@ -16,31 +16,25 @@ namespace boost { namespace simd
...
     value which will not overflow when multiplied by the input.
 
-    @par Semantic:
+   @par Note
```
Comment (jfalcou) [inline include/boost/simd/function/safe_max.hpp]: again, make it semantic or remove as it's a Impl. detail

File: include/boost/simd/function/saturate.hpp [hunk 33]
```diff
@@ -16,11 +16,8 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing saturate capabilities
-
-    Returns the saturated value of the first input in the template parameter type,
+    This function object returns the saturated value of the first input in the template parameter type,
```
Comment (jfalcou) [inline include/boost/simd/function/saturate.hpp:20]: input > argument

File: include/boost/simd/function/sign.hpp [hunk 34]
```diff
@@ -16,26 +16,25 @@ namespace boost { namespace simd
 {
 
  /*!
-
     @ingroup group-ieee
-    Function object implementing sign capabilities
-
-    Returns the sign of x. i.e. -1, 0 or 1, according
+    This function object returns the sign of x. i.e. -1, 0 or 1, according
```
Comment (jfalcou) [inline include/boost/simd/function/sign.hpp]: depending on the fact x is ...

File: include/boost/simd/function/ulpdist.hpp [hunk 35]
```diff
@@ -16,25 +16,8 @@ namespace boost { namespace simd
...
-    @code
-    T r = ulpdist(x,y);
-    @endcode
-
-    is similar to:
-
-    @code
-    T r = abs(x-y)/Eps<T>();
-    @endcode
+    This function object returns ulp distance of the two values.
```
Comment (jfalcou) [inline include/boost/simd/function/ulpdist.hpp]: the ulp distance between its arguments

**Comment by jfalcou** [comment]
@alankelly  Care to make a pass over this one ?

File: include/boost/simd/function/frexp.hpp (hunk 16 above)
Comment (jtlap) [reply include/boost/simd/function/frexp.hpp:20]: a mantissa

File: include/boost/simd/function/frexp.hpp (hunk 17 above)
Comment (jtlap) [reply include/boost/simd/function/frexp.hpp:24]: RealValue make the job I think

File: include/boost/simd/function/frexp.hpp (hunk 18 above)
Comment (jtlap) [reply include/boost/simd/function/frexp.hpp]: This has changed

File: include/boost/simd/function/frexp.hpp (hunk 16 above)
Comment (jfalcou) [reply include/boost/simd/function/frexp.hpp:20]: ok so which one "a manitssa" means we have a choice ?

File: include/boost/simd/function/frexp.hpp (hunk 17 above)
Comment (jfalcou) [reply include/boost/simd/function/frexp.hpp:24]: True

File: include/boost/simd/function/frexp.hpp (hunk 18 above)
Comment (jfalcou) [reply include/boost/simd/function/frexp.hpp]: the text or the behavior ?

File: include/boost/simd/function/negate.hpp (hunk 29 above)
Comment (jtlap) [reply include/boost/simd/function/negate.hpp]: there are 3 case -1, 0, 1 ... it seems verbose to me

File: include/boost/simd/function/next.hpp (hunk 30 above)
Comment (jtlap) [reply include/boost/simd/function/next.hpp]: they are stated above

File: include/boost/simd/function/maxmag.hpp [hunk 36]
```diff
@@ -17,7 +17,8 @@ namespace boost { namespace simd
 
  /*!
     @ingroup group-ieee
-    This function object returns the input value which have the greatest absolute value.
+    This function object returns the value with the greatest absolute value ammong
```
Comment (jfalcou) [inline include/boost/simd/function/maxmag.hpp]: among

File: include/boost/simd/function/maxnum.hpp [hunk 37]
```diff
@@ -17,7 +17,8 @@ namespace boost { namespace simd
 
  /*!
     @ingroup group-ieee
-    This function object returns the maximum value, ignoring nans.
+    This function object returns  the larger of two floating point arguments, treating NaNs
```
Comment (jfalcou) [inline include/boost/simd/function/maxnum.hpp]: the largest

File: include/boost/simd/function/next.hpp [hunk 38]
```diff
@@ -18,7 +18,7 @@ namespace boost { namespace simd
  /*!
     @ingroup group-ieee
     This function object returns the smallest representable value
-    strictly greater than the parameter with a few exceptions
+    strictly greater than the parameter with a few  exceptions or pecumliarities stated above:
```
Comment (jfalcou) [inline include/boost/simd/function/next.hpp]: there's nothing above :) 
Say: 

"strictly greater than its arguments except for : "

File: include/boost/simd/function/saturate.hpp [hunk 39]
```diff
@@ -18,7 +18,7 @@ namespace boost { namespace simd
  /*!
     @ingroup group-ieee
     This function object returns the saturated value of the first input in the template parameter type,
-    but in the same type as the input.
+    but in the same type as the input argument.
```
Comment (jfalcou) [inline include/boost/simd/function/saturate.hpp]: as its argument. It's obviously an input

File: include/boost/simd/function/bitfloating.hpp [hunk 40]
```diff
@@ -16,32 +16,32 @@ namespace boost { namespace simd
...
 
-    This function is the converse of @ref bitinteger
+    @par Note:
+    This  function is  related to the computation of
+    floating successors or predecessors as the ordering of the integers obtained
+    by @c bifloating is the same as the original order on the floating elements,
+    an increment can be used to find the successor.
 
-    @par Semantic:
+    `next(x)` is simply `bitfloating(inc(bitinteger(x)))`
```
Comment (jfalcou) [inline include/boost/simd/function/bitfloating.hpp]: I think this is also not required there

**Comment by jfalcou** [comment]
We still have some comments you didn't addres

File: include/boost/simd/function/frexp.hpp (hunk 16 above)
Comment (jtlap) [reply include/boost/simd/function/frexp.hpp:20]: Yes  a pair mantissa exponent has just to support m*2^e =x
and there is a choice in avx-512 for instance.
I wrote it but perhaps still not pushed

File: include/boost/simd/function/frexp.hpp (hunk 18 above)
Comment (jtlap) [reply include/boost/simd/function/frexp.hpp]: a mistake

File: include/boost/simd/function/frexp.hpp (hunk 19 above)
Comment (jtlap) [reply include/boost/simd/function/frexp.hpp:41]: that is done

Comment (jtlap) [reply include/boost/simd/function/frexp.hpp:41]: I do not understand why this does not disappear

**Pull request merged**